internal/gitprovider/github/github.go

```diff
@@ -0,0 +1,164 @@
+package github
+
+import (
+	"context"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/google/go-github/v56/github"
+	"github.com/pkg/errors"
+	"k8s.io/utils/pointer"
+
+	"github.com/akuity/kargo/internal/gitprovider"
+)
+
+var (
+	githubRegistration = gitprovider.ProviderRegistration{
+		Predicate: func(repoURL string) bool {
+			u, err := url.Parse(repoURL)
+			if err != nil {
+				return false
+			}
+			// The assumption here is that no one would ever put "github" in
+			// the DNS name of their self-hosted git provider. There is also
+			// the chance that they self host github but give it a name like:
+			// 'git.mycompany.com', in which case we would return false
```

Comment:
Isn't there a good possibility that someone self-hosting GitHub (e.g. GitHub Enterprise) _does_ have the word "github" in a subdomain?

Reply:
I think I need to revise this comment but basically it's trying explain that: if we see github in the hostname, we can safely assume that the git provider is github. If the word github **_doesn't_** appear in the hostname, we will return false, and not detect that github is the provider. The comment explains that someone might be hosting GitHub but not incorporate the word "github" into the name and then we would miss it.

> The assumption here is that no one would ever put "github" in the DNS name of their self-hosted git provider.

I'll rephrase this to say:
> We assume that any hostname with the word "github" in the hostname, can use this provider.

Reply:
Ah. Ok... now I understand what the intent of the comment was meant to be. 👍 

fwiw, I do I have some _vague_ recollection from previous projects about GitHub Enterprise being a somewhat different beasty API-wise from regular GitHub, but since I cannot recall specifics, that's something we can revisit at a later point if we determine its necessary.